compiler/rustc_trait_selection/src/solve/eval_ctxt.rs

```diff
@@ -164,6 +180,36 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
         let mode = if infcx.intercrate { SolverMode::Coherence } else { SolverMode::Normal };
         let mut search_graph = search_graph::SearchGraph::new(infcx.tcx, mode);
 
+        let inspect = {
+            let generate_proof_tree = match (
+                infcx.tcx.sess.opts.unstable_opts.dump_solver_proof_tree,
+                infcx.tcx.sess.opts.unstable_opts.dump_solver_proof_tree_uses_cache,
+                generate_proof_tree,
+            ) {
+                (_, Some(use_cache), GenerateProofTree::Yes(_)) => {
+                    GenerateProofTree::Yes(DisableGlobalCache::from_bool(!use_cache))
+                }
+
+                (SolverProofTreeCondition::Always, use_cache, GenerateProofTree::No) => {
+                    let use_cache = use_cache.unwrap_or(true);
+                    GenerateProofTree::Yes(DisableGlobalCache::from_bool(!use_cache))
+                }
+
+                (_, None, GenerateProofTree::Yes(_)) => generate_proof_tree,
+                // `Never` is kind of weird- it doesn't actually force us to not generate proof trees
```

Comment:
Maybe we should rename it from never then?

Reply:
yea